libraries/plugins/peerplays_sidechain/sidechain_net_handler.cpp

```diff
@@ -44,11 +44,42 @@ std::vector<std::string> sidechain_net_handler::get_sidechain_addresses() {
 void sidechain_net_handler::sidechain_event_data_received(const sidechain_event_data& sed) {
    ilog( __FUNCTION__ );
    ilog( "sidechain_event_data:" );
-   ilog( "  sidechain:      ${sidechain}", ( "sidechain", sed.sidechain ) );
-   ilog( "  transaction_id: ${transaction_id}", ( "transaction_id", sed.transaction_id ) );
-   ilog( "  from:           ${from}", ( "from", sed.from ) );
-   ilog( "  to:             ${to}", ( "to", sed.to ) );
-   ilog( "  amount:         ${amount}", ( "amount", sed.amount ) );
+   ilog( "  timestamp:                ${timestamp}", ( "timestamp", sed.timestamp ) );
+   ilog( "  sidechain:                ${sidechain}", ( "sidechain", sed.sidechain ) );
+   ilog( "  sidechain_uid:            ${uid}", ( "uid", sed.sidechain_uid ) );
+   ilog( "  sidechain_transaction_id: ${transaction_id}", ( "transaction_id", sed.sidechain_transaction_id ) );
+   ilog( "  sidechain_from:           ${from}", ( "from", sed.sidechain_from ) );
+   ilog( "  sidechain_to:             ${to}", ( "to", sed.sidechain_to ) );
+   ilog( "  sidechain_amount:         ${amount}", ( "amount", sed.sidechain_amount ) );
+   ilog( "  peerplays_from:           ${peerplays_from}", ( "peerplays_from", sed.peerplays_from ) );
+   ilog( "  peerplays_to:             ${peerplays_to}", ( "peerplays_to", sed.peerplays_to ) );
+
+   const chain::global_property_object& gpo = database.get_global_properties();
+
+   son_wallet_transfer_create_operation op;
+   op.payer = gpo.parameters.get_son_btc_account_id();
+   op.timestamp = sed.timestamp;
+   op.sidechain = sed.sidechain;
+   op.sidechain_uid = sed.sidechain_uid;
+   op.sidechain_transaction_id = sed.sidechain_transaction_id;
+   op.sidechain_from = sed.sidechain_from;
+   op.sidechain_to = sed.sidechain_to;
+   op.sidechain_amount = sed.sidechain_amount;
+   op.peerplays_from = sed.peerplays_from;
+   op.peerplays_to = sed.peerplays_to;
+
+   proposal_create_operation proposal_op;
+   proposal_op.fee_paying_account = plugin.get_son_object().son_account;
+   proposal_op.proposed_ops.push_back( op_wrapper( op ) );
+   uint32_t lifetime = ( gpo.parameters.block_interval * gpo.active_witnesses.size() ) * 3;
+   proposal_op.expiration_time = time_point_sec( database.head_block_time().sec_since_epoch() + lifetime );
+
+   signed_transaction trx = plugin.database().create_signed_transaction(plugin.get_private_keys().begin()->second, proposal_op);
+   try {
+      database.push_transaction(trx);
```

Comment:
We can add the check whether the SON is active one or not. However, this function executes independently of son scheduling, it is driven by events on Bitcoin network. As multiple SONs will detect Bitcoin event, independently of eachother, they will all raise proposals at more or less same time. They will not be able to detect duplicates, as the block containing this change is not yet generated and distributed between the nodes. Bottom line is that it is very hard to avoid creating duplicated proposals, but what we can do is to change the operation behaviour. Currently, if the transfer is already existing in the database, operation will fail. We can change this to "confirming" the existing transfer, so we will have the information how many SONs saw the same transaction on bitcoin network.

Reply:
What I meant was, 
"database.push_transaction(trx);"   pushes transaction into local database without publishing it to network.

If we are not an active witness and not producing blocks, will this transaction be sent out to all the nodes?

Reply:
I did not test that yet. We can add it if its needed.